### PR TITLE
[New arch] Migrate OCFile to new database

### PR DESCRIPTION
**Pull request by abelgardep**
Implements: #2955 
Needs: https://github.com/owncloud/android-library/pull/353

### QA

Test Plan: https://github.com/owncloud/QA/blob/master/Mobile/Android/Release_2.16/2995-DB_files_migration.md

Bugs & improvements:

- [x] (1) Install from scratch https://github.com/owncloud/android/pull/2995#issuecomment-723038851 [FIXED]
- [X] (2) Timestamps migrated in negative valies https://github.com/owncloud/android/pull/2995#issuecomment-724638539 [FIXED] 
- [x] (3) etag is not migrated for root folder https://github.com/owncloud/android/pull/2995#issuecomment-724719952 [NOT REPRODUCIBLE]

### CODE REVIEW

File: owncloudDomain/src/main/java/com/owncloud/android/domain/ext/AnyExt.kt
Comment (davigonz) [inline owncloudDomain/src/main/java/com/owncloud/android/domain/ext/AnyExt.kt:22]: ❤️

File: owncloudData/src/test/java/com/owncloud/android/data/file/datasources/OCLocalFileDataSourceTest.kt
Comment (davigonz) [inline owncloudData/src/test/java/com/owncloud/android/data/file/datasources/OCLocalFileDataSourceTest.kt:45]: I would add this as in a companion and uppercase with snake case `DUMMY_FILE_ENTITY`

File: owncloudData/src/main/java/com/owncloud/android/data/files/db/OCFileEntity.kt
Comment (davigonz) [inline owncloudData/src/main/java/com/owncloud/android/data/files/db/OCFileEntity.kt:69]: Add it as //TODO to check it in the future and not forget it?

File: owncloudData/src/main/java/com/owncloud/android/data/files/datasources/mapper/RemoteFileMapper.kt
Comment (davigonz) [inline owncloudData/src/main/java/com/owncloud/android/data/files/datasources/mapper/RemoteFileMapper.kt:32]: `if (it.isFolder) { it.size } else { it.length }` is easier to read imho

**Review by davigonz — CHANGES_REQUESTED** [review]

**Comment by davigonz** [comment]
@abelgardep just reviewed part of the PR

File: owncloudDomain/src/test/java/com/owncloud/android/domain/files/model/OCFileTest.kt
Comment (abelgardep) [reply owncloudDomain/src/test/java/com/owncloud/android/domain/files/model/OCFileTest.kt:21]: Yes, we have both since files are the most common datamodel in the app. This way we won't need to create a copy of OC_FILE each time we want a folder.

File: owncloudDomain/src/main/java/com/owncloud/android/domain/files/model/OCFile.kt
Comment (abelgardep) [reply owncloudDomain/src/main/java/com/owncloud/android/domain/files/model/OCFile.kt:48]: I separated must have and may have attributes. We will try to perform a cleanup. It will be fixed before merging new_arch/synchronization into master 👍

File: owncloudDomain/src/main/java/com/owncloud/android/domain/files/model/OCFile.kt
Comment (abelgardep) [reply owncloudDomain/src/main/java/com/owncloud/android/domain/files/model/OCFile.kt:64]: Because i need to create a new ocfile from java files. As soon as we move those files to kotlin, i will remove that constructor.

File: owncloudDomain/src/main/java/com/owncloud/android/domain/files/model/OCFile.kt
Comment (abelgardep) [reply owncloudDomain/src/main/java/com/owncloud/android/domain/files/model/OCFile.kt:111]: Yes 👍

**Comment by abelgardep** [comment]
> @abelgardep just reviewed part of the PR

Thanks! Changes applied!

File: owncloudDomain/src/test/java/com/owncloud/android/domain/files/model/OCFileTest.kt
Comment (davigonz) [reply owncloudDomain/src/test/java/com/owncloud/android/domain/files/model/OCFileTest.kt:21]: I mean calling it `DUMMY_OC_FILE` instead of `OC_FILE`...

File: owncloudDomain/src/main/java/com/owncloud/android/domain/files/model/OCFile.kt
Comment (davigonz) [reply owncloudDomain/src/main/java/com/owncloud/android/domain/files/model/OCFile.kt:64]: But can't you call the default constructor from java?

File: owncloudDomain/src/test/java/com/owncloud/android/domain/files/model/OCFileTest.kt
Comment (abelgardep) [reply owncloudDomain/src/test/java/com/owncloud/android/domain/files/model/OCFileTest.kt:21]: We call `OC_` instead of `DUMMY_`. We did that for capabilities and shares too.

File: owncloudDomain/src/main/java/com/owncloud/android/domain/files/model/OCFile.kt
Comment (abelgardep) [reply owncloudDomain/src/main/java/com/owncloud/android/domain/files/model/OCFile.kt:64]: No, because nullable and not nullable attributes are intercalated. I don't like to create a constructor for this, but the plan is to remove as soon as possible once we move forward more new_arch stuff. This will be merged to new_arch/synchronization and there is still a lot of work to do before merging it to master.

File: owncloudApp/src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.kt
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.kt:265]: Have you tried to solve this scenario without using `runBlocking`? I know the oC app is a special case because it has Java and Kotlin classes and we might need to call from Java to Kotlin. It's normal to use `runBlocking` in testing but not in the rest of the code because it blocks the thread (UI or not), so we lose the async approach...

File: owncloudApp/src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.kt
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.kt:316]: Please remove this or add a `//TODO` comment

File: owncloudApp/src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.kt
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.kt:399]: Same here?

File: owncloudApp/src/main/java/com/owncloud/android/datamodel/OCFile.java
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/datamodel/OCFile.java:46]: I would use 

`@Deprecated("Use com.owncloud.android.domain.files.model.OCFile instead")`

File: owncloudApp/src/main/java/com/owncloud/android/datamodel/OCUpload.java
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/datamodel/OCUpload.java:215]: This is a bit redundant, isn't it?

File: owncloudApp/src/main/java/com/owncloud/android/extensions/ViewModelExt.kt
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/extensions/ViewModelExt.kt:54]: Don't you do anything with this `useCaseResult`? How the user is notified about the usecase not being executed because of lack of connection?

File: owncloudApp/src/main/java/com/owncloud/android/operations/CreateFolderOperation.java
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/operations/CreateFolderOperation.java:38]: `@Deprecated("Call usecase instead of this operation. Remove as soon as possible")`

File: owncloudApp/src/main/java/com/owncloud/android/operations/RenameFileOperation.java
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/operations/RenameFileOperation.java:116]: Why? Add `//TODO` comment ?

File: owncloudApp/src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java:67]: Include comment inside `@Deprecated` ?

This way we can see more info when moving the mouse over the deprecated object, without having to enter this class.

File: owncloudApp/src/main/java/com/owncloud/android/presentation/viewmodels/files/FilesViewModel.kt
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/viewmodels/files/FilesViewModel.kt:35]: Very glad to see this viewmodel finally, awesome job 👏

File: owncloudApp/src/main/java/com/owncloud/android/presentation/viewmodels/files/FilesViewModel.kt
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/viewmodels/files/FilesViewModel.kt:33]: At this point we don't know what is under the hood of the use cases being called, it could be a repository or who knows...

File: owncloudApp/src/main/java/com/owncloud/android/providers/FileContentProvider.kt
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/FileContentProvider.kt:1075]: file list to new files table?

File: owncloudApp/src/main/java/com/owncloud/android/providers/FileContentProvider.kt
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/FileContentProvider.kt:1084]: Add a //TODO ?

File: owncloudApp/src/main/java/com/owncloud/android/ui/activity/FileActivity.java
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/ui/activity/FileActivity.java:496]: Passed as a parameter to what?

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoFragment.java
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoFragment.java:517]: Is this properly indented?

File: owncloudApp/src/main/java/com/owncloud/android/utils/UriUtilsKt.kt
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/utils/UriUtilsKt.kt:50]:
```suggestion
        return Uri.Builder().apply {
             scheme(ContentResolver.SCHEME_FILE)
             path(file.storagePath)
        }
```

Just a suggestion of course, your way is totally right as well 🍻

File: owncloudData/src/test/java/com/owncloud/android/data/file/datasources/OCLocalFileDataSourceTest.kt
Comment (davigonz) [inline owncloudData/src/test/java/com/owncloud/android/data/file/datasources/OCLocalFileDataSourceTest.kt:123]:
```suggestion
        verify { 
            dao.getFileByOwnerAndRemotePath(OC_FILE.owner, ROOT_PATH)
            dao.mergeRemoteAndLocalFile(any())
            dao.getFileById(1234)
        }
```

Can you try this?

**Review by davigonz — CHANGES_REQUESTED** [review]

File: owncloudApp/src/main/java/com/owncloud/android/datamodel/OCFile.java
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/datamodel/OCFile.java:46]: I wish i could, this is not kotlin :(

File: owncloudApp/src/main/java/com/owncloud/android/ui/activity/FileActivity.java
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/ui/activity/FileActivity.java:496]: 🤔  Im not sure. It was there. It was reordered by 'Reformat file'

File: owncloudApp/src/main/java/com/owncloud/android/operations/CreateFolderOperation.java
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/operations/CreateFolderOperation.java:38]: I cannot add reason in java files 😭

File: owncloudApp/src/main/java/com/owncloud/android/extensions/ViewModelExt.kt
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/extensions/ViewModelExt.kt:54]: Seems like i forgot to remove this function. Finally i used the `runUseCaseWithResult` one. Thanks

File: owncloudApp/src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java:67]: Adding reason, replaceWith etc, are options available only in kotlin. So i cannot use them in java world 😭

File: owncloudApp/src/main/java/com/owncloud/android/presentation/viewmodels/files/FilesViewModel.kt
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/viewmodels/files/FilesViewModel.kt:35]: This one will be a big one 🥳

File: owncloudApp/src/main/java/com/owncloud/android/providers/FileContentProvider.kt
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/providers/FileContentProvider.kt:1075]: If you know what i mean 😉

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoFragment.java
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoFragment.java:517]: Yes, it is

File: owncloudApp/src/main/java/com/owncloud/android/utils/UriUtilsKt.kt
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/utils/UriUtilsKt.kt:50]: Ok, applied.

```
        return Uri.Builder().apply {
            scheme(ContentResolver.SCHEME_FILE)
            path(file.storagePath)
        }.build()
```

File: owncloudData/src/test/java/com/owncloud/android/data/file/datasources/OCLocalFileDataSourceTest.kt
Comment (abelgardep) [reply owncloudData/src/test/java/com/owncloud/android/data/file/datasources/OCLocalFileDataSourceTest.kt:123]: Working fine too.

File: owncloudApp/src/main/java/com/owncloud/android/operations/RenameFileOperation.java
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/operations/RenameFileOperation.java:116]: Added Fixme.

File: owncloudApp/src/main/java/com/owncloud/android/ui/activity/FileActivity.java
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/ui/activity/FileActivity.java:496]: It was there before, it was reordered by 'Reformat file' 🤔

File: owncloudApp/src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.kt
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.kt]: The idea is to remove FileDataStorageManager and call those usecases from FilesViewModel. I have added some todos, and added IO as runblocking context. This is an intermediate step, it won't be that way in the final version.

File: owncloudApp/src/main/java/com/owncloud/android/datamodel/OCFile.java
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/datamodel/OCFile.java:46]: tenor

File: owncloudApp/src/main/java/com/owncloud/android/operations/CreateFolderOperation.java
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/operations/CreateFolderOperation.java:38]: :sad:

File: owncloudApp/src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java:67]: Yep, sorry

**Review by davigonz — APPROVED** [review]

**Comment by jesmrec** [comment]
### (1)

1. Install the app from scratch
2. Enter URL of basic auth and credentials

Current: app runs into ANR. After closing/opening 2 or 3 times, the list of files is finally displayed. 
Expected: Smooth login

Pixel 2, Android 11
`65ebec7`

**Comment by abelgardep** [comment]
> ### (1)
> 1. Install the app from scratch
> 2. Enter URL of basic auth and credentials
> 
> Current: app runs into ANR. After closing/opening 2 or 3 times, the list of files is finally displayed.
> Expected: Smooth login
> 
> Pixel 2, Android 11
> `65ebec7`

We got blocked by the account manager. We will retrieve the ownCloud client dynamically in a safe thread using the client manager. This should fix the problem. Ready to test @jesmrec

**Comment by jesmrec** [comment]
### (2)

1. Install `master` branch
2. Get data from database, paying some attention to fields `last_sync_date` and  `last_sync_date_for_data`
3. Upgrade to current branch
4. Check fields `last_sync_date` and  `lastSyncDateForData` and `lastSyncDateForProperties`

Those fields contain negative values.

This is before upgrading: 

Screenshot 2020-11-10 at 12 10 54


And after upgrading:

Screenshot 2020-11-10 at 12 16 25

How i did upgrade: with Android Studio by checking out the current branch and the option `Run ownCloud App`

Pixel2 Android11
`112807a`

**Comment by abelgardep** [comment]
> ### (2)
> 1. Install `master` branch
> 2. Get data from database, paying some attention to fields `last_sync_date` and  `last_sync_date_for_data`
> 3. Upgrade to current branch
> 4. Check fields `last_sync_date` and  `lastSyncDateForData` and `lastSyncDateForProperties`
> 
> Those fields contain negative values.
> 

Good catch, fixed! Ready to test @jesmrec 👍

**Comment by jesmrec** [comment]
### (3)

1. Install `master` branch
2. Get data from database, paying some attention to field `etag` of `/`entry (`id` = 1) -> contains value like `5fa4f98f94522`
3. Upgrade to current branch
3. Check field `etag` for the `/`  entry (`id` = 1)

Current: empty
Expected: contains the same value as in `master`

Pixel2 Android11
`112807a`

**Comment by jesmrec** [comment]
Approved on my side

**Pull request merged**
